core/src/main/java/io/jenkins/pluginhealth/scoring/probes/SCMLinkValidationProbe.java

```diff
@@ -42,20 +53,94 @@
 @Component
 @Order(value = SCMLinkValidationProbe.ORDER)
 public class SCMLinkValidationProbe extends Probe {
-    private static final Logger LOGGER = LoggerFactory.getLogger(SCMLinkValidationProbe.class);
-
-    private static final String GH_REGEXP = "https://(?<server>[^/]*)/(?<repo>jenkinsci/[^/]*)(?:/(?<folder>.*))?";
-    public static final Pattern GH_PATTERN = Pattern.compile(GH_REGEXP);
     public static final int ORDER = UpdateCenterPluginPublicationProbe.ORDER + 100;
     public static final String KEY = "scm";
+    private static final Logger LOGGER = LoggerFactory.getLogger(SCMLinkValidationProbe.class);
+    private static final String GH_REGEXP = "https://(?<server>[^/]*)/(?<repo>jenkinsci/[^/]*)";
+    public static final Pattern GH_PATTERN = Pattern.compile(GH_REGEXP);
 
     @Override
     public ProbeResult doApply(Plugin plugin, ProbeContext context) {
         if (plugin.getScm() == null || plugin.getScm().isBlank()) {
             LOGGER.warn("{} has no SCM link", plugin.getName());
-            return ProbeResult.error(key(), "The plugin SCM link is empty");
+            return ProbeResult.error(key(), "The plugin SCM link is empty.");
+        }
+        return fromSCMLink(context, plugin.getScm(), plugin.getName());
+    }
+
+    /**
+     * Validates the SCM link, and sets {@link ProbeContext#setScmFolderPath(Optional)}. The value is always the path of the POM file.
+     *
+     * @param context    Refer {@link ProbeContext}.
+     * @param scm        The SCM link {@link Plugin#getScm()}.
+     * @param pluginName The name of the plugin {@link Plugin#getName()}.
+     * @return ProbeResult {@link ProbeResult}.
+     */
+    private ProbeResult fromSCMLink(ProbeContext context, String scm, String pluginName) {
+        Matcher matcher = GH_PATTERN.matcher(scm);
+        if (!matcher.find()) {
+            LOGGER.atDebug().log(() -> String.format("%s is not respecting the SCM URL Template.", scm));
+            return ProbeResult.failure(key(), "SCM link doesn't match GitHub plugin repositories.");
+        }
+        try {
+            context.getGitHub().getRepository(matcher.group("repo"));
+            Optional<Path> pluginPathInRepository = findPluginPom(context.getScmRepository(), pluginName);
+            Optional<Path> folderPath = pluginPathInRepository.map(path -> path.getParent());
+            if (folderPath.isEmpty()) {
+                return ProbeResult.error(key(), String.format("No valid POM file found in %s plugin.", pluginName));
+            }
+            context.setScmFolderPath(Optional.of(folderPath.get().getFileName()));
```

Comment:
Here I believe it would be better to have the ScmFolderPath being a `Optional<String>`.

Considering `folderPath` is already an `Optional`, you can simply go for something like:

`context.setScmFolderPath(folderPath.map( whatever function gives you the value you expected in there ));`

If the `Optional` was empty, the map function won't be running, and you'll store an empty again. Otherwise, the Optional will contain a `String` with the value you wanted.

Reply:
Looks good the way you wrote it in [6f90c9b](https://github.com/jenkins-infra/plugin-health-scoring/pull/351/commits/6f90c9b255cf7f26faafec3dae12d253e17f4eed)